src/uct/ib/mlx5/ib_mlx5.h

```diff
@@ -247,13 +247,43 @@ typedef union {
 } uct_ib_mlx5_devx_mr_t;
 
 
+/* Data structure to hold the UMR MR (on the host) item in the mkey pool */
+typedef struct {
+    ucs_list_link_t    super;
+    struct mlx5dv_mkey *umr_mkey;
+} uct_ib_mlx5_devx_umr_t;
+
+
+#define UCT_IB_MLX5_UMR_FMT "UMR mkey %p index 0x%x"
+#define UCT_IB_MLX5_UMR_ARG(umr) \
+    (umr)->umr_mkey, uct_ib_mlx5_mkey_index((umr)->umr_mkey->lkey)
+
+
+/* Data structure to hold the UMR mkey alias on the DPU item in the hash map */
+typedef struct {
+    struct mlx5dv_devx_obj *cross_mr;
+    uint32_t               lkey;
+} uct_ib_mlx5_devx_umr_alias_t;
+
+
+#define UCT_IB_MLX5_UMR_ALIAS_FMT "UMR mkey alias %p index 0x%x"
+#define UCT_IB_MLX5_UMR_ALIAS_ARG(umr_alias) \
+    (umr_alias)->cross_mr, uct_ib_mlx5_mkey_index((umr_alias)->lkey)
+
+
+/* Hash map of indirect mkey (from the host) to mkey alias (on the DPU) */
+/* Note the hash key here is: gvmi_id << 32 | mkey (both uint32_t) */
+KHASH_MAP_INIT_INT64(umr_mkey_map, uct_ib_mlx5_devx_umr_alias_t);
+
+
 typedef struct {
     uct_ib_mem_t            super;
     void                    *address;
     struct mlx5dv_devx_obj  *atomic_dvmr;
     struct mlx5dv_devx_obj  *indirect_dvmr;
     struct mlx5dv_devx_umem *umem;
     struct mlx5dv_devx_obj  *cross_mr;
+    uct_ib_mlx5_devx_umr_t  *umr_mr;
```

Comment:
maybe it should be called exported_umr_mkey?

Reply:
sounds good, renamed